File: src/ZoneCodeGenerator/Domain/Evaluation/OperationType.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;

namespace ZoneCodeGenerator.Domain.Evaluation
{
    sealed class OperationType
    {
        // https://en.cppreference.com/w/cpp/language/operator_precedence
        public enum OperationPrecedence
        {
            MultiplicationDivisionRemainder = 1,
            AdditionSubtraction = 2,
            BitwiseShift = 3,
            RelationalGreaterLessThan = 4,
            RelationalEquals = 5,
            LogicalAnd = 6,
            LogicalOr = 7
        }

        public delegate int EvaluationFunction(int operand1, int operand2);

        public string Syntax { get; }
        public OperationPrecedence Precedence { get; }
        public EvaluationFunction Function { get; }

        public override string ToString()
        {
            return Syntax;
        }

        private OperationType(string syntax, OperationPrecedence precedence, EvaluationFunction function)
        {
            Syntax = syntax;
            Precedence = precedence;
            Function = function;
        }

        public static OperationType OperationAdd = new OperationType(
            "+",
            OperationPrecedence.AdditionSubtraction,
            (operand1, operand2) => operand1 + operand2
            );

        public static OperationType OperationSubtract = new OperationType(
            "-",
            OperationPrecedence.AdditionSubtraction,
            (operand1, operand2) => operand1 - operand2
            );

        public static OperationType OperationMultiply = new OperationType(
            "*",
            OperationPrecedence.MultiplicationDivisionRemainder,
            (operand1, operand2) => operand1 * operand2
            );

        public static OperationType OperationDivide = new OperationType(
            "/",
            OperationPrecedence.MultiplicationDivisionRemainder,
            (operand1, operand2) => operand1 / operand2
            );

        public static OperationType OperationRemainder = new OperationType(
            "%",
            OperationPrecedence.MultiplicationDivisionRemainder,
            (operand1, operand2) => operand1 % operand2
            );

        public static OperationType OperationShiftLeft = new OperationType(
            "<<",
            OperationPrecedence.BitwiseShift,
            (operand1, operand2) => operand1 << operand2
            );

        public static OperationType OperationShiftRight = new OperationType(
            ">>",
            OperationPrecedence.BitwiseShift,
            (operand1, operand2) => operand1 >> operand2
            );

        public static OperationType OperationGreaterThan = new OperationType(
            ">",
            OperationPrecedence.RelationalGreaterLessThan,
            (operand1, operand2) => operand1 > operand2 ? 1 : 0
            );

        public static OperationType OperationGreaterEqualsThan = new OperationType(
            ">=",
            OperationPrecedence.RelationalGreaterLessThan,
            (operand1, operand2) => operand1 >= operand2 ? 1 : 0
            );

        public static OperationType OperationLessThan = new OperationType(
            "<",
            OperationPrecedence.RelationalGreaterLessThan,
            (operand1, operand2) => operand1 < operand2 ? 1 : 0
            );

        public static OperationType OperationLessEqualsThan = new OperationType(
            "<=",
            OperationPrecedence.RelationalGreaterLessThan,
            (operand1, operand2) => operand1 <= operand2 ? 1 : 0
            );

        public static OperationType OperationEquals = new OperationType(
            "==",
            OperationPrecedence.RelationalEquals,
            (operand1, operand2) => operand1 == operand2 ? 1 : 0
            );

        public static OperationType OperationNotEquals = new OperationType(
            "!=",
            OperationPrecedence.RelationalEquals,
            (operand1, operand2) => operand1 != operand2 ? 1 : 0
            );

        public static OperationType OperationAnd = new OperationType(
            "&&",
            OperationPrecedence.LogicalAnd,
            (operand1, operand2) => operand1 > 0 && operand2 > 0 ? 1 : 0
            );

        public static OperationType OperationOr = new OperationType(
            "||",
            OperationPrecedence.LogicalOr,
            (operand1, operand2) => operand1 > 0 || operand2 > 0 ? 1 : 0
            );

        public static List<OperationType> Types => typeof(OperationType)
            .GetFields(BindingFlags.Static | BindingFlags.Public)
            .Select(info => info.GetValue(null))
            .OfType<OperationType>()
            .ToList();
    }
}

File: src/ZoneCodeGeneratorLib/Domain/Evaluation/OperationType.cpp
#include "OperationType.h"

OperationType::OperationType(std::string syntax, const OperationPrecedence precedence, std::function<int(int, int)> evaluationFunction)
    : m_syntax(std::move(syntax)),
      m_precedence(precedence),
      m_evaluation_function(std::move(evaluationFunction))
{
}

const OperationType* const OperationType::OPERATION_ADD
    = new OperationType("+", OperationPrecedence::ADDITION_SUBTRACTION, [](auto op1, auto op2)
    {
        return op1 + op2;
    });

const OperationType* const OperationType::OPERATION_SUBTRACT
    = new OperationType("-", OperationPrecedence::ADDITION_SUBTRACTION, [](auto op1, auto op2)
    {
        return op1 - op2;
    });

const OperationType* const OperationType::OPERATION_MULTIPLY
    = new OperationType("*", OperationPrecedence::MULTIPLICATION_DIVISION_REMAINDER, [](auto op1, auto op2)
    {
        return op1 * op2;
    });

const OperationType* const OperationType::OPERATION_DIVIDE
    = new OperationType("/", OperationPrecedence::MULTIPLICATION_DIVISION_REMAINDER, [](auto op1, auto op2)
    {
        return op1 / op2;
    });

const OperationType* const OperationType::OPERATION_REMAINDER
    = new OperationType("%", OperationPrecedence::MULTIPLICATION_DIVISION_REMAINDER, [](auto op1, auto op2)
    {
        return op1 % op2;
    });

const OperationType* const OperationType::OPERATION_BITWISE_AND
    = new OperationType("&", OperationPrecedence::BITWISE_AND, [](auto op1, auto op2)
    {
        return op1 & op2;
    });

const OperationType* const OperationType::OPERATION_BITWISE_XOR
    = new OperationType("^", OperationPrecedence::BITWISE_XOR, [](auto op1, auto op2)
    {
        return op1 ^ op2;
    });

const OperationType* const OperationType::OPERATION_BITWISE_OR
    = new OperationType("|", OperationPrecedence::BITWISE_OR, [](auto op1, auto op2)
    {
        return op1 | op2;
    });

const OperationType* const OperationType::OPERATION_SHIFT_LEFT
    = new OperationType("<<", OperationPrecedence::BITWISE_SHIFT, [](auto op1, auto op2)
    {
        return op1 << op2;
    });

const OperationType* const OperationType::OPERATION_SHIFT_RIGHT
    = new OperationType(">>", OperationPrecedence::BITWISE_SHIFT, [](auto op1, auto op2)
    {
        return op1 >> op2;
    });

const OperationType* const OperationType::OPERATION_GREATER_THAN
    = new OperationType(">", OperationPrecedence::RELATIONAL_GREATER_LESS_THAN, [](auto op1, auto op2)
    {
        return op1 > op2 ? 1 : 0;
    });

const OperationType* const OperationType::OPERATION_GREATER_EQUAL_THAN
    = new OperationType(">=", OperationPrecedence::RELATIONAL_GREATER_LESS_THAN, [](auto op1, auto op2)
    {
        return op1 >= op2 ? 1 : 0;
    });

const OperationType* const OperationType::OPERATION_LESS_THAN
    = new OperationType("<", OperationPrecedence::RELATIONAL_GREATER_LESS_THAN, [](auto op1, auto op2)
    {
        return op1 < op2 ? 1 : 0;
    });

const OperationType* const OperationType::OPERATION_LESS_EQUAL_THAN
    = new OperationType("<=", OperationPrecedence::RELATIONAL_GREATER_LESS_THAN, [](auto op1, auto op2)
    {
        return op1 <= op2 ? 1 : 0;
    });

const OperationType* const OperationType::OPERATION_EQUALS
    = new OperationType("==", OperationPrecedence::RELATIONAL_EQUALS, [](auto op1, auto op2)
    {
        return op1 == op2 ? 1 : 0;
    });

const OperationType* const OperationType::OPERATION_NOT_EQUAL
    = new OperationType("!=", OperationPrecedence::RELATIONAL_EQUALS, [](auto op1, auto op2)
    {
        return op1 != op2 ? 1 : 0;
    });

const OperationType* const OperationType::OPERATION_AND
    = new OperationType("&&", OperationPrecedence::LOGICAL_AND, [](auto op1, auto op2)
    {
        return op1 > 0 && op2 > 0 ? 1 : 0;
    });

const OperationType* const OperationType::OPERATION_OR
    = new OperationType("||", OperationPrecedence::LOGICAL_OR, [](auto op1, auto op2)
    {
        return op1 > 0 || op2 > 0 ? 1 : 0;
    });


const OperationType* const OperationType::ALL_OPERATION_TYPES[]
{
    OPERATION_ADD,
    OPERATION_SUBTRACT,
    OPERATION_MULTIPLY,
    OPERATION_DIVIDE,
    OPERATION_REMAINDER,
    OPERATION_BITWISE_AND,
    OPERATION_BITWISE_XOR,
    OPERATION_BITWISE_OR,
    OPERATION_SHIFT_LEFT,
    OPERATION_SHIFT_RIGHT,
    OPERATION_GREATER_THAN,
    OPERATION_GREATER_EQUAL_THAN,
    OPERATION_LESS_THAN,
    OPERATION_LESS_EQUAL_THAN,
    OPERATION_EQUALS,
    OPERATION_NOT_EQUAL,
    OPERATION_AND,
    OPERATION_OR
};
